File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode[]} lists
# @return {ListNode}
def merge_k_lists(lists)
    vals=[]
    lists.each do |list| 
        while list
            vals.push(list.val)
            list=list.next
        end
    end
    return vals if vals.empty?
    vals.sort!
    head=ListNode.new(vals[0])
    iterator=head
    vals.slice!(0,1)
    vals.each do |val|
        iterator.next=ListNode.new(val)
        iterator=iterator.next
    end
    head
# ---------------------------------------------------------------

#     iterator,sorted=nil,nil
#     lists.filter! {|node| node}
    
#     while !lists.empty?
#         min_index=0
#        lists.each_with_index do |node,index|
#            min_index=index if node.val<lists[min_index].val
#        end
#         new_node=ListNode.new(lists[min_index].val)
#         lists[min_index]=lists[min_index].next
#         lists.slice!(min_index,1) if lists[min_index].nil?
#         if sorted.nil?
#             sorted=new_node 
#             iterator=sorted
#         else
#             iterator.next=new_node
#             iterator=iterator.next
#         end
#     end
  
# ---------------------------------------------------------------
   #     def recurse nodes
   #     # p nodes
   #     return nil if nodes.length==0
   #     return nodes[0] if nodes.length==1
   #     max_index=0
   #     nodes.each_with_index do |node,index|
   #         max_index=index if node.val<nodes[max_index].val
   #     end
   #     new_node=ListNode.new(nodes[max_index].val)
   #     nodes[max_index]=nodes[max_index].next
   #      nodes.slice!(max_index,1) if nodes[max_index].nil?
   #     next_node=recurse(nodes)
   #     new_node.next=next_node
   #     new_node
   # end
   #  recurse lists
    # sorted
end

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
const merge = function(n1, n2){
    if(!n1||!n2)
        return n1?n1:n2
    let iterator=null
    let temp=n1
    let tempN2=n2
    if(n1.val<=n2.val){
        iterator=n1
        temp=n1.next
    }else{
        iterator=n2
        tempN2=n2.next
    }
    let head=iterator
    while(temp){
        while(tempN2&&tempN2.val<=temp.val){
            iterator.next=tempN2
            tempN2=tempN2.next
            iterator=iterator.next
        }
        iterator.next=temp
        temp=temp.next
        iterator=iterator.next
    }
    iterator.next=null
    if(tempN2){
        iterator.next=tempN2
    }
    return head
}
var mergeKLists = function(lists) {
    // consider the first linked list is sorted 
        // iteratvely merge the rest of the linked list to the first list 
        // complexity O(n*m)
    if(lists.length<1)
        return null
    let head=lists[0]
    for(let i=1; i<lists.length; i++){
        head=merge(head, lists[i])
    }
    return head;
}; 